src/lib/PWM/PWM_ESP32.cpp

```diff
@@ -0,0 +1,222 @@
+#define LEDC_CHANNELS   8
+
+#define MCPWM_CHANNEL_FLAG  0x100
+#define LEDC_CHANNEL_FLAG   0x200
+
+#define IS_MCPWM_CHANNEL(ch) (ch & MCPWM_CHANNEL_FLAG)
+#define IS_LEDC_CHANNEL(ch) (ch & LEDC_CHANNEL_FLAG)
+
+#define LEDC_CHANNEL(ch) (ch & 0xFF)
+#define MCPWM_CHANNEL(ch) (ch & 0xFF)
+
+static const struct {
+    mcpwm_unit_t unit;
+    mcpwm_io_signals_t signal;
+    mcpwm_timer_t timer;
+    mcpwm_generator_t generator;
+} mcpwm_config[] = {
+    {MCPWM_UNIT_0, MCPWM0A, MCPWM_TIMER_0, MCPWM_GEN_A},
+    {MCPWM_UNIT_0, MCPWM0B, MCPWM_TIMER_0, MCPWM_GEN_B},
+    {MCPWM_UNIT_0, MCPWM1A, MCPWM_TIMER_1, MCPWM_GEN_A},
+    {MCPWM_UNIT_0, MCPWM1B, MCPWM_TIMER_1, MCPWM_GEN_B},
+    {MCPWM_UNIT_0, MCPWM2A, MCPWM_TIMER_2, MCPWM_GEN_A},
+    {MCPWM_UNIT_0, MCPWM2B, MCPWM_TIMER_2, MCPWM_GEN_B},
+    {MCPWM_UNIT_1, MCPWM0A, MCPWM_TIMER_0, MCPWM_GEN_A},
+    {MCPWM_UNIT_1, MCPWM0B, MCPWM_TIMER_0, MCPWM_GEN_B},
+    {MCPWM_UNIT_1, MCPWM1A, MCPWM_TIMER_1, MCPWM_GEN_A},
+    {MCPWM_UNIT_1, MCPWM1B, MCPWM_TIMER_1, MCPWM_GEN_B},
+    {MCPWM_UNIT_1, MCPWM2A, MCPWM_TIMER_2, MCPWM_GEN_A},
+    {MCPWM_UNIT_1, MCPWM2B, MCPWM_TIMER_2, MCPWM_GEN_B},
+};
+
+static uint32_t mcpwm_frequencies[MCPWM_CHANNELS] = {0};
+
+static struct {
+    int8_t pin;
+    uint8_t resolution_bits;
+    uint32_t interval;
+} ledc_config[LEDC_CHANNELS] = {
+    {-1, 0, 0}
+};
+uint32_t ledcTimerConfigs[LEDC_TIMER_MAX] = {0};
+
+/*
+ * Modified versions of the ledcSetup/ledcAttachPin from Arduino ESP32 Hal which allows
+ * control of the timer to use rather than being directly associated with the channel. This
+ * allows multiple channels to use the same timer within a LEDC group (high/low speed) when
+ * they are set to the same frequency.
+ */
+static void ledcSetupEx(uint8_t chan, ledc_timer_t timer, uint32_t freq, uint8_t bit_num)
+{
+    ledc_timer_config_t ledc_timer = {
+        .speed_mode = LEDC_LOW_SPEED_MODE,
+        .duty_resolution = (ledc_timer_bit_t)bit_num,
+        .timer_num = timer,
+        .freq_hz = freq,
+        .clk_cfg = LEDC_USE_APB_CLK,
+    };
+    if (ledc_timer_config(&ledc_timer) != ESP_OK)
+    {
+        ERRLN("ledc setup failed!");
+    }
+}
+
+static void ledcAttachPinEx(uint8_t pin, uint8_t chan, ledc_timer_t timer)
+{
+    ledc_channel_config_t ledc_channel = {
+        .gpio_num = pin,
+        .speed_mode = LEDC_LOW_SPEED_MODE,
+        .channel = (ledc_channel_t)chan,
+        .intr_type = LEDC_INTR_DISABLE,
+        .timer_sel = (ledc_timer_t)timer,
+        .duty = 0,
+        .hpoint = 0};
+    ledc_channel_config(&ledc_channel);
+}
+
+pwm_channel_t PWMController::allocate(uint8_t pin, uint32_t frequency)
+{
+    // 1. see if we can allocate a MCPWM channel at this frequency
+    int channel = -1;
+    // 1a. see if theres a MCPWM already using this frequency we can piggy-back on
+    for (int i=0 ; i<MCPWM_CHANNELS ; i++)
+    {
+        if (mcpwm_frequencies[i] == frequency)
+        {
+            if (i%2 == 0 && mcpwm_frequencies[i+1]==0)
```

Comment:
The inconsistency WRT whitespace between the `==` is mildly triggering 😅

Reply:
I ran the code formatter so it's fixed all the similar problems too.